src/base/bittorrent/session.cpp

```diff
@@ -3324,6 +3324,8 @@ void Session::handleTorrentSavePathChanged(TorrentHandle *const torrent)
 void Session::handleTorrentCategoryChanged(TorrentHandle *const torrent, const QString &oldCategory)
 {
     saveTorrentResumeData(torrent);
+    // `torrentCategoryChanged` parameter `oldCategory` cannot be a const reference,
+    // as it might point to the updated data and not the old one we wanted.
```

Comment:
You're overreacting here... Could you explain, please?

Reply:
Of course not... changes in `TorrentHandle::setCategory` are not sufficient on its own. `CategoryFilterModel::torrentCategoryChanged` will get the updated category instead of the old one.
I admit I'm not so familiar with the internals of signal-slots so I can only guess the signal is emitted a bit later (not at the exact `emit`).

Reply:
>I admit I'm not so familiar with the internals of signal-slots so I can only guess the signal is queued and emitted a bit later (not at the exact `emit`).

Qt docs:
>When a signal is emitted, the slots connected to it are *usually executed immediately*, just like a normal function call. [...] Execution of the code following the `emit` statement will occur once all slots have returned.

So there is no problem!
```c++
const QString oldCategory = m_category; // store old category
m_category = category; // actually change category of torrent
m_session->handleTorrentCategoryChanged(this, oldCategory);
// observers can obtain both old (via slot parameter) 
// and new category (via `TorrentHandle::category()`)
```
Qt docs:
>The situation is slightly different when using `queued connections`; in such a case, the code following the emit keyword will continue immediately, and the slots will be executed later.

But here is also no problem even in case of "Queued connection". Qt copies the arguments to store them in an event behind the scenes. Otherwise, we could not be able to use slots with parameters of reference type in queued connections.

Reply:
>So there is no problem!

Right! But once I change it to https://github.com/qbittorrent/qBittorrent/pull/10816/commits/e0116b2239d1b2b4d3efe09a5e6338c0afdcf368#diff-cf66938d43d8a44584487ed5b6df6027R1279
It doesn't work as expected, is my code invalid?
You can dump the value `oldCategory` in `CategoryFilterModel::torrentCategoryChanged()` to verify it.

Reply:
Well, that will take time. Since everything else is OK, you can extract this commit into a separate PR.

Reply:
>Right! But once I change it to e0116b2#diff-cf66938d43d8a44584487ed5b6df6027R1279
It doesn't work as expected

I've made a quick review. I see no point with your changes. What do you expect from it?

Reply:
This line looks like a workaround to me:
https://github.com/qbittorrent/qBittorrent/commit/e0116b2239d1b2b4d3efe09a5e6338c0afdcf368#diff-cf66938d43d8a44584487ed5b6df6027L1279
Why assign the value to a temp variable and pass the reference when you can just pass the value out by making a copy?

Reply:
How do you suppose we should report old category name **after** it is changed without preliminary saving it in another variable?

Reply:
By making the signal to take a copy of it: https://github.com/qbittorrent/qBittorrent/commit/e0116b2239d1b2b4d3efe09a5e6338c0afdcf368#diff-b369e532eff50253387c9d203da4b928R461

Reply:
Sorry, @Chocobo1, looks like you didn't get much sleep... but you're not answering **my** question. Please read it again and pay attention to the **bold** part. Maybe then you'll understand...

Reply:
I realized (not coming from your question, sorry) that the new code invoked `m_session->handleTorrentCategoryChanged(this, m_category)` too early. At the time it invokes, the category hasn't really been changed and thus it is wrong behavior.
I'll drop the commit.